File: trees/binary-tree-inorder/binary-tree-inorder.ts
// Trees: Inorder traversal of a binary tree

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val: number, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function inorder(root: TreeNode | null, result: number[]): void {
  if (root === null) return;
  inorder(root.left, result);
  result.push(root.val);
  inorder(root.right, result);
}

function main(): void {
  const root = new TreeNode(2, new TreeNode(1), new TreeNode(3));

  const result: number[] = [];
  inorder(root, result);
  console.log(result);
}

main();

File: trees/binary-tree-inorder/binary-tree-inorder.java
// Trees: Inorder traversal of a binary tree
import java.util.ArrayList;
import java.util.List;

class BinaryTreeInorder {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val) { this.val = val; }
    }

    static void inorder(TreeNode root, List<Integer> result) {
        if (root == null) return;
        inorder(root.left, result);
        result.add(root.val);
        inorder(root.right, result);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);

        List<Integer> result = new ArrayList<>();
        inorder(root, result);

        System.out.println(result);
    }
}
